hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala

```diff
@@ -0,0 +1,388 @@
+          statusList += fileStatus(blobFileName)
+        }
+      }
+    }
+    populateStatList()
+
+    statusList.toArray
+  }
+
+  def glob(filename: String): Array[FileStatus] = {
+    var (account, container, path) = getAccountContainerPath(filename)
+    path = dropTrailingSlash(path)
+
+    val components =
+      if (path == "")
+        Array.empty[String]
+      else
+        path.split("/")
+
+    val javaFS = FileSystems.getDefault
+
+    val ab = new mutable.ArrayBuffer[FileStatus]()
+    def f(prefix: String, fs: FileStatus, i: Int): Unit = {
+      assert(!prefix.endsWith("/"), prefix)
+
+      if (i == components.length) {
+        var t = fs
+        if (t == null) {
+          try {
+            t = fileStatus(prefix)
+          } catch {
+            case _: FileNotFoundException =>
+          }
+        }
+        if (t != null)
+          ab += t
+      }
+
+      if (i < components.length) {
+        val c = components(i)
+        if (containsWildcard(c)) {
+          val m = javaFS.getPathMatcher(s"glob:$c")
+          for (cfs <- listStatus(prefix)) {
+            val p = dropTrailingSlash(cfs.getPath)
+            val d = p.drop(prefix.length + 1)
+            if (m.matches(javaFS.getPath(d))) {
+              f(p, cfs, i + 1)
+            }
+          }
+        } else
+          f(s"$prefix/$c", null, i + 1)
+      }
+    }
+
+    f(s"hail-az://$account/$container", null, 0)
+    ab.toArray
+  }
+
+  def fileStatus(filename: String): FileStatus = {
+    val (account, container, path) = getAccountContainerPath(filename)
+
+    if (path == "") {
+      return new BlobStorageFileStatus(s"hail-az://$account/$container", null, 0, true)
+    }
+
+    val blobClient: BlobClient = getBlobClient(filename)
+    val blobContainerClient: BlobContainerClient = getContainerClient(filename)
+
+    val directoryContents = blobContainerClient.listBlobsByHierarchy(dropTrailingSlash(path) + "/")
+    var seenBlobItems: Set[String] = Set()
+    var numSubBlobs: Int = 0
+    val iter = directoryContents.iterator()
+    def countSubBlobs(): Unit = {
+      while (iter.hasNext) {
+        val name = iter.next.getName
+        if (seenBlobItems.contains(name)) return
+        seenBlobItems += name
+        numSubBlobs += 1
+      }
+    }
+    countSubBlobs()
+
+    if (numSubBlobs == 0 && !blobClient.exists()) throw new FileNotFoundException("file isn't found")
+
+    val isDir = numSubBlobs > 0
+    if (isDir) {
+      new BlobStorageFileStatus(dropTrailingSlash(filename), -1, -1, isDir)
+    }
+    else {
+      val blobProperties: BlobProperties = blobClient.getProperties
+      val modificationTime = blobProperties.getLastModified.toEpochSecond
+      val size = blobProperties.getBlobSize
+
+      new BlobStorageFileStatus(dropTrailingSlash(filename), modificationTime, size, isDir)
+    }
+  }
+
+  def makeQualified(filename: String): String = {
+    assert(filename.startsWith("hail-az://"))
```

Comment:
Same here with raising a `ValueError`?

Reply:
We should raise `ValueError` here instead so that clients have an opportunity to `catch` it and recover instead of the program crashing.

Reply:
That would be a good test to add actually

Reply:
Gotcha, I can add that test. I'm also going to change this behavior in the GoogleStorageFS